File: src/mainApp/GameObject.java
package mainApp;
/**
 * Class: mainApp
 * @author Team 303
 * <br>Purpose: Super class that parents all possible objects that will be on the screen
 * <br>Restrictions: Needs x and y coordinates to be passed in that are valid for the screen
 * <br>For example: 
 * <pre>
 *    GameObject g = new GameObject(10, 10);
 * </pre>
 */
import java.awt.Graphics2D;
import java.awt.geom.Rectangle2D;

// TODO: Make abstract?

/**
 * Class: GameObject
 * @author Team 303
 * <br>Purpose: Parent class for every object in the game (Platforms, the player, enemies, etc.)
 */
public class GameObject {
	private double xCent;
	private double yCent;
	private double Width;
	private double Height;
	private boolean shouldRemove;
	
	public GameObject(double xCord, double yCord) {
			this.xCent = xCord;
			this.yCent = yCord;
			this.Width = 50;
			this.Height = 50;
			this.shouldRemove = false;
	}

	public double getWidth() {
		return Width;
	}

	public void setWidth(double width) {
		Width = width;
	}

	public double getHeight() {
		return Height;
	}

	public void setHeight(double height) {
		Height = height;
	}

	public void update() throws DeadException {
		
	}
	
	public void drawOn(Graphics2D g2) {
		// TODO Auto-generated method stub
		
	}

	public double getXCent() {
		return xCent;
	}

	public void setXCent(double d) {
		this.xCent = d;
	}

	public double getYCent() {
		return yCent;
	}

	public void setYCent(double yCent) {
		this.yCent = yCent;
	}
	public Rectangle2D.Double getBoundingBox() {
		return new Rectangle2D.Double(this.xCent - this.getWidth()/2, this.yCent - this.getHeight()/2, this.getWidth(), this.getHeight() );
	}
	public boolean overlaps(GameObject other) {
		return getBoundingBox().intersects(other.getBoundingBox());
	}
	
	public void markForRemoval() {
		this.shouldRemove = true;
	}
	
	public boolean shouldBeRemoved() {
		return this.shouldRemove;
	}
	
	public double getDistance(double x, double y) {
		return Math.sqrt(Math.pow(x - this.getXCent(), 2) + Math.pow(y - this.getYCent(), 2));
	}

}

File: docs/js/GameObject.js
/**
 * Base class for all game objects
 */
class GameObject {
    constructor(xCord, yCord) {
        this.xCent = xCord;
        this.yCent = yCord;
        this.width = 50;
        this.height = 50;
        this.shouldRemove = false;
    }

    getWidth() {
        return this.width;
    }

    setWidth(width) {
        this.width = width;
    }

    getHeight() {
        return this.height;
    }

    setHeight(height) {
        this.height = height;
    }

    update() {
        // Override in subclasses
    }

    drawOn(ctx) {
        // Override in subclasses
    }

    getXCent() {
        return this.xCent;
    }

    setXCent(x) {
        this.xCent = x;
    }

    getYCent() {
        return this.yCent;
    }

    setYCent(y) {
        this.yCent = y;
    }

    getBoundingBox() {
        return {
            x: this.xCent - this.width / 2,
            y: this.yCent - this.height / 2,
            width: this.width,
            height: this.height
        };
    }

    overlaps(other) {
        const rect1 = this.getBoundingBox();
        const rect2 = other.getBoundingBox();
        
        const isOverlapping = rect1.x < rect2.x + rect2.width &&
               rect1.x + rect1.width > rect2.x &&
               rect1.y < rect2.y + rect2.height &&
               rect1.y + rect1.height > rect2.y;
        
        // Debug: Log critical overlaps only
        if (this.constructor.name === 'Hero' && other.constructor.name === 'Platform' && isOverlapping && this.firstCollision !== true) {
            console.log(`First collision detected: Hero overlaps Platform`);
            this.firstCollision = true;
        }
        
        return isOverlapping;
    }

    markForRemoval() {
        this.shouldRemove = true;
    }

    shouldBeRemoved() {
        return this.shouldRemove;
    }

    getDistance(x, y) {
        return Math.sqrt(Math.pow(x - this.xCent, 2) + Math.pow(y - this.yCent, 2));
    }
} 